### PR TITLE
Bluetooth: Mesh: Introduce separate workq for ADV EXT

### PR DESCRIPTION
**Pull request by LingaoM**
Using an independent Mesh workq will better ensure the reliability of mesh message sending, including when sysworkq is blocked elsewhere, also to make the host always send packets

### CODE REVIEW

File: subsys/bluetooth/mesh/adv_ext.c [hunk 1]
```diff
@@ -493,26 +510,37 @@ int bt_mesh_adv_enable(void)
...
 		atomic_set_bit(advs[i].flags, ADV_FLAG_SUSPENDING);
 
-		if (k_current_get() != &k_sys_work_q.thread ||
-		    (k_work_busy_get(&advs[i].work) & K_WORK_RUNNING) == 0) {
+		if (!(k_work_busy_get(&advs[i].work) & K_WORK_RUNNING)
+#if defined(CONFIG_BT_MESH_WORKQ_SYS)
+		    || (k_current_get() != &k_sys_work_q.thread)) {
+#else  /* CONFIG_BT_MESH_WORKQ_MESH */
+		    || (k_current_get() != &bt_mesh_workq.thread)) {
+#endif /* CONFIG_BT_MESH_WORKQ_SYS */
```
Comment (jhedberg) [inline subsys/bluetooth/mesh/adv_ext.c:536]: I think this could be made slightly more readable with something like:
```c
#if defined(CONFIG_BT_MESH_WORKQ_SYS)
#define MESH_WORKQ_THREAD &k_sys_work_q.thread
#else
#define MESH_WORKQ_THREAD &bt_mesh_workq.thread
#endif
...
	if (!k_work_busy_get(&advs[i].work) & K_WORK_RUNNING) ||
	    k_current_get() != MESH_WORKQ_THREAD) {
		k_work_flush(&advs[i].work, &sync);
	}
```

**Review by jhedberg — DISMISSED** [review]

**Review by jhedberg — DISMISSED** [review]

**Comment by alxelax** [comment]
Hi, considering issues with collaboration between Mesh and Host do we really need to use system work queue?
https://github.com/zephyrproject-rtos/zephyr/issues/77241
Probably it is worth to leave only specific queue that @LingaoM added here and remove the system work queue usage.
@PavelVPV, @jhedberg opinions?

**Comment by LingaoM** [comment]
> Hi, considering issues with collaboration between Mesh and Host do we really need to use system work queue? #77241 Probably it is worth to leave only specific queue that @LingaoM added here and remove the system work queue usage. @PavelVPV, @jhedberg opinions?

I looked back at commit history, the first time sending Mesh messages use extended advertising was used in sysworkq was introduced : https://github.com/zephyrproject-rtos/zephyr/pull/30005 .
At this time, the Host send_sync(will handle by BT TX thread), so there are no risk for Conflict or dead-lock.
But after https://github.com/zephyrproject-rtos/zephyr/pull/72090, BT TX thread was removed, and will increase risk deak-lock with Zephyr Bluetooth Mesh implementation, which leads to the controversy here:https://github.com/zephyrproject-rtos/zephyr/issues/77241

So, from my point of view, this commit should be submitted as a PR with the commit to reconstruct BT TX.

**Review by PavelVPV — CHANGES_REQUESTED** [review]
> But after #72090, BT TX thread was removed, also use sysworkq, thus will increase risk deak-lock with Zephyr Bluetooth Mesh implementation, which leads to the controversy here:#77241

From my current understanding, the deadlock won't happen because of this PR: https://github.com/zephyrproject-rtos/zephyr/pull/71697.

Also, this is not the only place where the stack uses sysworkq to send messages. Transport SAR and beacons when sending to proxy client or server do it from sysworkq as well. But the same workaround is done in the host to avoid deadlocks, e.g.:https://github.com/zephyrproject-rtos/zephyr/blob/2eda5f3e10f653525a79c18e5acfbc7877dced42/subsys/bluetooth/host/conn.c#L1582-L1586

So currently, from my understanding, no deadlocks should happen. The consequence of this is that a PDU will simply not be sent.

If, however, the intention of this PR is to make the host always send packets, then other places in the mesh stack should be considered.

**Comment by PavelVPV** [comment]
@LingaoM are you still working on the PR? I started reviewing it but I see you did some changes.

**Comment by LingaoM** [comment]
> @LingaoM are you still working on the PR? I started reviewing it but I see you did some changes.

Done, now, thanks.

**Review by PavelVPV — CHANGES_REQUESTED** [review]
I wonder if we try another approach and try to back-off when the host API returns an error which per my understanding (or please correct me if I missing something) can now happen if the host ran out of cmd buffers. If this worked then we wouldn't need an extra thread and could just try to reschedule the advertiser and send the same adv again (same for proxy).

One problem I could imaging with this is that if some other code also tries to send something from a higher priority context, then mesh can never get a change to send anything. But this I think can be checked.

File: subsys/bluetooth/mesh/adv_legacy.c [hunk 2]
```diff
@@ -258,3 +258,8 @@ int bt_mesh_adv_gatt_start(const struct bt_le_adv_param *param, int32_t duration
 	adv_timeout = duration;
 	return bt_le_adv_start(param, ad, ad_len, sd, sd_len);
 }
+
+int bt_mesh_wq_submit(struct k_work *work)
+{
+	return k_work_submit(work);
```
[the remark this answers is not in the document]
Comment (LingaoM) [reply subsys/bluetooth/mesh/adv_legacy.c:264]: Unfortunately, legacy advertiser is not a workq type.

File: subsys/bluetooth/mesh/proxy_msg.c [hunk 3]
```diff
@@ -230,6 +241,46 @@ int bt_mesh_proxy_relay_send(struct bt_conn *conn, struct bt_mesh_adv *adv)
...
+		struct bt_mesh_proxy_role *role = &roles[bt_conn_index(conn)];
+
+		k_fifo_put(&role->pending, bt_mesh_adv_ref(adv));
+		bt_mesh_wq_submit(&role->work);
+
+		return 0;
+	}
+#endif /* CONFIG_BT_MESH_PROXY_MSG_ALWAYS_SEND */
+
+	return proxy_relay_send(conn, adv);
```
[the remark this answers is not in the document]
Comment (LingaoM) [reply subsys/bluetooth/mesh/proxy_msg.c:257]: I have also considered this, but I think **relay** messages are not in order, because from the perspective of radio propagation, there is a possibility that the actual message sent later may be received first because the actual message sent earlier is interfered with.

File: subsys/bluetooth/mesh/adv_ext.c [hunk 4]
```diff
@@ -424,6 +441,13 @@ void bt_mesh_adv_init(void)
 	for (int i = 0; i < ARRAY_SIZE(advs); i++) {
 		(void)memcpy(&advs[i].adv_param, &adv_param, sizeof(adv_param));
 	}
+
+#if defined(CONFIG_BT_MESH_WORKQ_MESH)
```
[the remark this answers is not in the document]
Comment (LingaoM) [reply subsys/bluetooth/mesh/adv_ext.c:445]: image

File: tests/bsim/bluetooth/mesh/tests_scripts/proxy_sol/sol_replay.sh [hunk 5]
```diff
@@ -32,6 +32,18 @@ RunTest mesh_srpl_replay_attack \
 	proxy_sol_iut_power_replay_attack \
 	-flash=../results/mesh_srpl_replay_attack/flash.bin -flash_rm
 
+overlay="overlay_pst_conf_overlay_gatt_conf_overlay_workq_mesh_conf"
+RunTest mesh_srpl_replay_attack \
+	proxy_sol_tester_immediate_replay_attack \
```
[the remark this answers is not in the document]
Comment (LingaoM) [reply tests/bsim/bluetooth/mesh/tests_scripts/proxy_sol/sol_replay.sh:37]: I guess you mean `mesh_srpl_replay_attack_workq`.

File: subsys/bluetooth/mesh/adv_ext.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (PavelVPV) [reply subsys/bluetooth/mesh/adv_ext.c:445]: Yes, the `#if defined(CONFIG_BT_MESH_WORKQ_MESH)` above should be removed I guess. I wonder if that leaves any traces behind in the firmware.

[the remark this answers is not in the document]
Comment (LingaoM) [reply subsys/bluetooth/mesh/adv_ext.c:436]: I suggest keeping the status quo so that the code will be more straightforward to read, especially for newcomers.

File: subsys/bluetooth/mesh/proxy_msg.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (LingaoM) [reply subsys/bluetooth/mesh/proxy_msg.c]: > I wonder if we try another approach and try to back-off when the host API returns an error which per my understanding (or please correct me if I missing something) can now happen if the host ran out of cmd buffers. If this worked then we wouldn't need an extra thread and could just try to reschedule the advertiser and send the same adv again (same for proxy).

It has been modified based on this, so the issue of out-of-order becomes even less important.

File: subsys/bluetooth/mesh/Kconfig [hunk 6]
```diff
@@ -103,6 +112,57 @@ menuconfig BT_MESH_ADV_EXT
 
 if BT_MESH_ADV_EXT
 
+choice BT_MESH_WORKQ_CONTEXT
```
Comment (alxelax) [inline subsys/bluetooth/mesh/Kconfig:115]: I still do not understand why `system queue` is required if there is a separate work queue for it.
I didn't get any meaningful explanation of it. What is use case for it?
It looks like unnecessary overcomplication to me, both implementation and maintenance.

Comment (LingaoM) [reply subsys/bluetooth/mesh/Kconfig:115]: I'm thinks the bigger reason of keep sysworkq is less ram comsume.

File: subsys/bluetooth/mesh/adv_ext.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (PavelVPV) [reply subsys/bluetooth/mesh/adv_ext.c:436]: https://github.com/zephyrproject-rtos/zephyr/blob/e40fa8914fade81b91c6ef4f1e618ca5d9adf444/subsys/bluetooth/mesh/settings.c#L52-L59

https://github.com/zephyrproject-rtos/zephyr/blob/e40fa8914fade81b91c6ef4f1e618ca5d9adf444/subsys/bluetooth/mesh/settings.c#L272-L277

File: subsys/bluetooth/mesh/Kconfig (hunk 6 above)
Comment (LingaoM) [reply subsys/bluetooth/mesh/Kconfig:115]: https://github.com/zephyrproject-rtos/zephyr/actions/runs/11288115630/job/31395274065?pr=79690

I have to say that using K_NO_WAIT is not as stable as before.

Comment (LingaoM) [reply subsys/bluetooth/mesh/Kconfig:115]: I agree with your opinion. I have resubmitted it now and set Mesh Workq as the default, and added a description in the doc. 

1. For bsim, all tests will be based on Mesh Workq by default, and additional tests based on sysworkq will be added.

3. For resource overhead, I think it is not easy to use adv_ext.c for products with insufficient resources, such as microbit_v1.

Comment (LingaoM) [reply subsys/bluetooth/mesh/Kconfig:115]: As for whether to retain sysworkq, I think we can refer to the implementation of hci core(https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/hci_core.c#L4072). Maybe we can remove it later, but I think there is no big problem in retaining it now.

**Comment by PavelVPV** [comment]
> I wonder if we try another approach and try to back-off when the host API returns an error which per my understanding (or please correct me if I missing something) can now happen if the host ran out of cmd buffers. If this worked then we wouldn't need an extra thread and could just try to reschedule the advertiser and send the same adv again (same for proxy).
> 
> One problem I could imaging with this is that if some other code also tries to send something from a higher priority context, then mesh can never get a change to send anything. But this I think can be checked.

Answering to myself. I implemented a stress test to check this: https://github.com/PavelVPV/zephyr/commit/15c9049f34a99fd0520ccb7384676b4a335a724a

Yes, this approach will work, but will put a lot of strain on CPU because the host doesn't have an API to tell that it has a free buffer. So for now using separate thread will guarantee that stack will always be able to send a message.

File: subsys/bluetooth/mesh/adv_ext.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (LingaoM) [reply subsys/bluetooth/mesh/adv_ext.c]: Done.

File: tests/bsim/bluetooth/mesh/overlay_workq_mesh.conf
Comment (PavelVPV) [inline tests/bsim/bluetooth/mesh/overlay_workq_mesh.conf:1]: Rename to `overlay_workq_sys.conf`? It is a bit confusing as it looks like enables `CONFIG_BT_MESH_WORKQ_MESH`.

File: subsys/bluetooth/mesh/Kconfig (hunk 6 above)
Comment (PavelVPV) [reply subsys/bluetooth/mesh/Kconfig:115]: I agree that we need to keep sysworkq option for devices with the limited RAM. IF the mesh stack is the only user of the host, it is possible to estimate the required number of tx buffers, which will always work.

File: subsys/bluetooth/mesh/Kconfig [hunk 7]
```diff
@@ -103,6 +112,58 @@ menuconfig BT_MESH_ADV_EXT
...
+config BT_MESH_WORKQ_SYS
+	bool "Sending for BT Mesh messages in the system work queue"
+	help
+	  When this option is selected, the Mesh will sending for Mesh messages
+	  in the system work queue. The application needs to ensure the system
+	  workqueue stack size (SYSTEM_WORKQUEUE_STACK_SIZE) is large enough,
+	  refer to BT_MESH_ADV_STACK_SIZE for the recommended minimum.
+	  Note: When this option is used, other users of the system work queue will
+	  influence the latency of sending Mesh messages, and the reliability of sending
+	  Mesh messages will be decreased.
```
Comment (PavelVPV) [inline subsys/bluetooth/mesh/Kconfig:163]:
```suggestion
	bool "Send messages from the system workqueue"
	help
	  When this option is selected, the mesh sends messages from
	  the system work queue. The application needs to ensure the system
	  workqueue stack size (SYSTEM_WORKQUEUE_STACK_SIZE) is large enough,
	  refer to BT_MESH_ADV_STACK_SIZE for the recommended minimum.
	  
	  When this option is enabled and the mesh tries to send a message,
	  and the host ran out the HCI command buffers controlled by
	  CONFIG_BT_BUF_CMD_TX_COUNT, the host returns -ENOBUFS immediately
	  and the mesh drops the message transmission. To mitigate this
	  issue, make sure to have sufficient number of HCI command buffers.

	  When this option is enabled, the latency of sending mesh messages
	  will be affected by other users on the system work queue, resulting in
	  reduced reliability for sending mesh messages.
```

File: subsys/bluetooth/mesh/Kconfig [hunk 8]
```diff
@@ -103,6 +112,58 @@ menuconfig BT_MESH_ADV_EXT
...
+	help
+	  Selects in which context the BT Mesh messages will be sending.
+
+config BT_MESH_WORKQ_MESH
+	bool "Sending for BT Mesh messages in the mesh-specific work queue"
+	help
+	  When this option is selected, the Mesh will sending for Mesh messages
+	  in the mesh-specific work queue. This will ensure Mesh messages always sent.
+	  The application needs to ensure the mesh-specific work queue size is large
+	  enough, refer to BT_MESH_ADV_STACK_SIZE for the recommended minimum.
```
Comment (PavelVPV) [inline subsys/bluetooth/mesh/Kconfig:127]:
```suggestion
	bool "Send messages from in the mesh-specific workqueue"
	help
	  When this option is selected, the mesh sends messages from the
	  mesh-specific workqueue. This will ensure that messages are always sent.

	  The application needs to ensure the mesh-specific workqueue size is large
	  enough. Refer to BT_MESH_ADV_STACK_SIZE for the recommended minimum.
```

File: subsys/bluetooth/mesh/Kconfig [hunk 9]
```diff
@@ -103,6 +112,58 @@ menuconfig BT_MESH_ADV_EXT
 
 if BT_MESH_ADV_EXT
 
+choice BT_MESH_WORKQ_CONTEXT
+	prompt "Mesh Thread Selection"
+	default BT_MESH_WORKQ_MESH
+	help
+	  Selects in which context the BT Mesh messages will be sending.
```
Comment (PavelVPV) [inline subsys/bluetooth/mesh/Kconfig:119]:
```suggestion
	prompt "Advertising thread selection"
	default BT_MESH_WORKQ_MESH
	help
	  Defines a context for mesh messages transmission.
```

File: doc/releases/release-notes-4.0.rst [hunk 10]
```diff
@@ -106,6 +106,12 @@ Bluetooth
 
 * HCI Drivers
 
+* Mesh
+
+  * Introduce Mesh Specific workq thread, with default enable. The main impact on users is that RAM
+    overhead takes up 1-2KB more. But the reliability of Mesh message sending will be significantly
+    increased.
```
Comment (PavelVPV) [inline doc/releases/release-notes-4.0.rst:113]:
```suggestion
  * Introduced a mesh-specific workqueue to increase reliability of the mesh messages 
    transmission. To get the old behavior enable :kconfig:option:`CONFIG_BT_MESH_WORKQ_SYS`.
```

File: subsys/bluetooth/mesh/Kconfig [hunk 11]
```diff
@@ -21,6 +21,15 @@ if BT_MESH
 config BT_MESH_PROXY
 	bool
 
+config BT_MESH_PROXY_MSG_ALWAYS_SEND
```
[the remark this answers is not in the document]
Comment (LingaoM) [reply subsys/bluetooth/mesh/Kconfig:24]: the new workq stack size will decrease if not enable this.

**Comment by LingaoM** [comment]
@PavelVPV bbsim also failed, I'm no ideas for this.

File: subsys/bluetooth/mesh/Kconfig (hunk 11 above)
[the remark this answers is not in the document]
Comment (PavelVPV) [reply subsys/bluetooth/mesh/Kconfig:24]: Yes, but it is not guaranteed that a message will be sent and then the mesh workq needs the bigger stack size.

It also creates re-ordering on the node which is not good.

**Comment by PavelVPV** [comment]
> @PavelVPV bbsim also failed, I'm no ideas for this.

The stack suspends faster than manages to complete publication. This should work:
```
diff --git a/tests/bsim/bluetooth/mesh/src/test_suspend.c b/tests/bsim/bluetooth/mesh/src/test_suspend.c
index 5ec6a8dc685..404ed4213df 100644
--- a/tests/bsim/bluetooth/mesh/src/test_suspend.c
+++ b/tests/bsim/bluetooth/mesh/src/test_suspend.c
@@ -295,6 +295,7 @@ static void dut_pub_common(bool disable_bt)
                ASSERT_OK_MSG(k_sem_take(&publish_sem, K_SECONDS(30)), "Pub timed out");
        }
 
+       k_sleep(K_MSEC(100));
        ASSERT_OK_MSG(bt_mesh_suspend(), "Failed to suspend Mesh.");
```

File: subsys/bluetooth/mesh/Kconfig (hunk 6 above)
Comment (LingaoM) [reply subsys/bluetooth/mesh/Kconfig:115]: I'm sorry, there was an issue with my previous BSIM environment. I will resolve it instead of bothering you.

File: subsys/bluetooth/mesh/Kconfig [hunk 12]
```diff
@@ -103,6 +103,64 @@ menuconfig BT_MESH_ADV_EXT
...
+	  When this option is selected, the mesh sends messages from the
+	  mesh-specific workqueue. This will ensure that messages are always sent.
+	  The application needs to ensure the mesh-specific workqueue size is large
+	  enough. Refer to BT_MESH_ADV_STACK_SIZE for the recommended minimum.
+
+if BT_MESH_WORKQ_MESH
+
+config BT_MESH_ADV_STACK_SIZE
+	int "Mesh extended advertiser thread stack size"
+	default 2048 if BT_MESH_PROXY
```
Comment (PavelVPV) [inline subsys/bluetooth/mesh/Kconfig:124]: Have you checked if this thread actually requires such big stack size?

File: subsys/bluetooth/mesh/adv_ext.c [hunk 13]
```diff
@@ -503,14 +527,13 @@ int bt_mesh_adv_enable(void)
...
-	int err;
 	struct k_work_sync sync;
+	int err;
 
 	for (int i = 0; i < ARRAY_SIZE(advs); i++) {
 		atomic_set_bit(advs[i].flags, ADV_FLAG_SUSPENDING);
 
-		if (k_current_get() != &k_sys_work_q.thread ||
-		    (k_work_busy_get(&advs[i].work) & K_WORK_RUNNING) == 0) {
+		if (k_current_get() != k_work_queue_thread_get(MESH_WORKQ)) {
```
Comment (PavelVPV) [inline subsys/bluetooth/mesh/adv_ext.c:536]: Why did you remove `(k_work_busy_get(&advs[i].work) & K_WORK_RUNNING) == 0)` check? It is needed to ensure that we are not calling `k_work_flush` from the advertiser's k_work handler. Otherwise it will deadlock.

**Review by PavelVPV — DISMISSED** [review]

File: doc/releases/release-notes-4.0.rst [hunk 14]
```diff
@@ -132,6 +132,11 @@ Bluetooth
 
 * HCI Drivers
 
+* Mesh
+
+  * Introduced a mesh-specific workqueue to increase reliability of the mesh messages
+    transmission. To get the old behavior enable :kconfig:option:`CONFIG_BT_MESH_WORKQ_SYS`.
```
[the remark this answers is not in the document]
Comment (PavelVPV) [reply doc/releases/release-notes-4.0.rst:138]: This partially fixes https://github.com/zephyrproject-rtos/zephyr/issues/77241 but I guess we can't use "Partially fixes #...." so probably a new issue should be created. By partially I mean we leave option with sysworkq and that may still not work in certain conditions. Also, ideally, this shouldn't be fixed by a new thread, but currently there is no alternative. So actual fix should included some sort of asynchronous API.

File: tests/bsim/bluetooth/mesh/tests_scripts/access/access_cancel.sh [hunk 15]
```diff
@@ -7,6 +7,10 @@ source $(dirname "${BASH_SOURCE[0]}")/../../_mesh_test.sh
 RunTest mesh_access_publication_cancel \
 	access_tx_cancel access_rx_cancel
 
+overlay=overlay_workq_sys_conf
```
Comment (alxelax) [inline tests/bsim/bluetooth/mesh/tests_scripts/access/access_cancel.sh:10]: Why was it added in access layer test? How does it impact on access?

**Review by alxelax — CHANGES_REQUESTED** [review]
LGTM, but it seems `proxy_mixin.sh` advertiser bsim test was missed with new overlay.
I guess it should be extended too.

**Review by PavelVPV — DISMISSED** [review]

**Review by alxelax — DISMISSED** [review]

File: subsys/bluetooth/mesh/Kconfig [hunk 16]
```diff
@@ -103,6 +103,64 @@ menuconfig BT_MESH_ADV_EXT
...
 if BT_MESH_ADV_EXT
 
+choice BT_MESH_WORKQ_CONTEXT
+	prompt "Advertising thread selection"
+	default BT_MESH_WORKQ_MESH
+	help
+	  Defines a context for mesh messages transmission.
+
+config BT_MESH_WORKQ_MESH
+	bool "Send messages from in the mesh-specific workqueue"
```
Comment (tejlmand) [inline subsys/bluetooth/mesh/Kconfig:113]: Consider reducing this title to:
```suggestion
	bool "Mesh-specific workqueue"
```

This will thing look neater in menuconfig, which is where the title has a role.
Without change:
image

With change:
image

File: subsys/bluetooth/mesh/Kconfig [hunk 17]
```diff
@@ -103,6 +103,64 @@ menuconfig BT_MESH_ADV_EXT
...
+	help
+	  When this option is selected, the mesh sends messages from the
+	  mesh-specific workqueue. This will ensure that messages are always sent.
+	  The application needs to ensure the mesh-specific workqueue size is large
+	  enough. Refer to BT_MESH_ADV_STACK_SIZE for the recommended minimum.
+
+if BT_MESH_WORKQ_MESH
+
+config BT_MESH_ADV_STACK_SIZE
```
Comment (tejlmand) [inline subsys/bluetooth/mesh/Kconfig:122]: It's confusing to have config options together with choice entries like this, as it makes it a harder for the developer to identify the specific choice entries.

image

Also note, that this construct is only possible because the config types are not bools.
Please move the int entries outside the choice.

File: subsys/bluetooth/mesh/Kconfig [hunk 18]
```diff
@@ -103,6 +103,64 @@ menuconfig BT_MESH_ADV_EXT
...
+	help
+	  Priority of bt mesh adv thread.
+
+	  NOTE: This is an advanced setting and should not be changed unless
+	  absolutely necessary
+
+endif # BT_MESH_WORKQ_MESH
+
+config BT_MESH_WORKQ_SYS
+	bool "Send messages from the system workqueue"
```
Comment (tejlmand) [inline subsys/bluetooth/mesh/Kconfig:146]: Related to other comment, consider reducing this title to:
```suggestion
	bool "System workqueue"
```

**Review by tejlmand — CHANGES_REQUESTED** [review]
Generally looks good, only some minor nits.

**Review by alxelax — APPROVED** [review]

**Review by tejlmand — APPROVED** [review]

**Review by PavelVPV — APPROVED** [review]
